File: test/net/sslserver.py
# coding: utf-8
import os, sys
import socket
import ssl

# create cert.pem by "openssl req -new -x509 -days 365 -nodes -out cert.pem -keyout cert.pem"

def test():
    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM, 0)
    sock = ssl.wrap_socket(sock,
                    server_side=True,
                    certfile="cert.pem",
                    keyfile=None,)
                    #ssl_version=ssl.PROTOCOL_TLSv1)

    sock.bind(('127.0.0.1', 10000))
    sock.listen(32)

    while True:
        newsock, addr = sock.accept()
        
        while True:
            ret = newsock.recv(1024)
            print 'read:', ret
            if not ret:
                print 'close'
                break
            if ret.startswith('quit'):
                print 'quit'
                break
            s = 'haha\r\n'
            print 'write:', s, newsock.send(s)

        newsock.close()

if __name__ == '__main__':
    test()

File: test/net/sslserver.c
#include <zocle/zocle.h>

int main()
{
    zc_mem_init(ZC_MEM_GLIBC|ZC_MEM_DBG_OVERFLOW);
    zc_log_new("stdout", ZC_LOG_ALL);

#ifdef ZOCLE_WITH_SSL
    zcSocket    *sock;

    zc_socket_startup();

    sock = zc_socket_server_tcp("127.0.0.1", 10000, 32);
    if (NULL == sock) {
        ZCERROR("tcp server create error!\n");
        return -1;
    }

    ZCINFO("tcp server start ok!\n");
    int ret;
    while (1) {
        zcSocket *newsock = zc_socket_accept(sock);
        if (NULL == newsock)
            break;

        ret = zc_socket_server_ssl(newsock, "cert.pem", "cert.pem", ZC_SSL_CERT_NONE, ZC_SSL_VER_SSL3, NULL);
        if (ret != ZC_OK) {
            ZCERROR("tcp ssl error! %d\n", ret);
            return -1;
        }

        char buf[1024];
        ZCINFO("new socket remote:%s:%d, local:%s:%d\n", 
                    newsock->remote.ip, newsock->remote.port,
                    newsock->local.ip, newsock->local.port);
        char sendbuf[128];
        while (1) {
            zc_check(newsock);
            ret = zc_socket_recv(newsock, buf, 1024);
            if (ret == 0) {
                ZCINFO("read 0, close conn.\n");
                //zc_socket_delete(newsock);
                break;
            }
            buf[ret] = 0;

            ZCINFO("recv %d: %s\n", ret, buf);
            sprintf(sendbuf, "good!\r\n");
            ret = zc_socket_send(newsock, sendbuf, strlen(sendbuf));
            if (ret != strlen(sendbuf)) {
                ZCINFO("send error:%d", ret);
                break;
            }
            ZCINFO("send:%s", sendbuf);
        }
        zc_check(newsock);
        zc_socket_delete(newsock);
    }
    
    zc_socket_delete(sock);
    zc_socket_cleanup();
#endif

    return 0;
}
